transform-plugins/src/main/java/io/cdap/plugin/XMLParser.java

```diff
@@ -77,66 +82,79 @@ public XMLParser(Config config) {
   @Override
   public void configurePipeline(PipelineConfigurer pipelineConfigurer) throws IllegalArgumentException {
     super.configurePipeline(pipelineConfigurer);
-    outSchema = config.getOutputSchema();
+    // Get failure collector for updated validation API
+    FailureCollector collector = pipelineConfigurer.getStageConfigurer().getFailureCollector();
+    outSchema = config.getOutputSchema(collector);
+    if (!this.config.containsMacro(INPUT)
+        && pipelineConfigurer.getStageConfigurer().getInputSchema().getField(this.config.inputField) == null) {
+      collector.addFailure(String.format("Input field '%s' must exist in input schema.", this.config.inputField), null)
+          .withConfigProperty(INPUT);
+    }
     try {
-      validateXpathAndSchema();
+      validateXpathAndSchema(collector);
     } catch (UnsupportedEncodingException e) {
-      throw new IllegalArgumentException("Failed to parse XPath mappings. Reason : " + e.getMessage());
+      collector.addFailure(String.format("Failed to parse XPath mappings: %s.", e.getMessage()), null)
+          .withConfigProperty(XPATH_MAPPINGS);
     }
     pipelineConfigurer.getStageConfigurer().setOutputSchema(outSchema);
   }
 
   @Override
   public void initialize(TransformContext context) throws Exception {
     super.initialize(context);
-    outSchema = config.getOutputSchema();
-    xPathMapping = getXPathMapping();
+    // Get failure collector for updated validation API
+    FailureCollector collector = getContext().getFailureCollector();
+    outSchema = config.getOutputSchema(collector);
+    xPathMapping = getXPathMapping(collector);
```

Comment:
call collector.getOrThrowException after this to fail the pipeline run here instead of continuing on.

This should actually be validated in prepareRun(), which is one once at the start of the run. initialize() is called by each mapper at the start of their lifecycle, so if there are 100 mappers created for the run, you'll see 100 error messages in the log, all saying the same thing

Reply:
Fixed, and moved the runtime validation to prepareRun instead of initialize

Reply:
Currently, the unit tests rely on initialize to create a mock transform context. For now I'll leave the current initialization logic as it is, but we should probably revisit it at some point.